backend/integrations/gitlab/gitlab.go

```diff
@@ -0,0 +1,365 @@
+	References           References           `json:"references"`
+	Severity             string               `json:"severity"`
+	Subscribed           bool                 `json:"subscribed"`
+	MovedToID            *int                 `json:"moved_to_id"`
+	ServiceDeskReplyTo   *string              `json:"service_desk_reply_to"`
+}
+
+type Author struct {
+	ID        int    `json:"id"`
+	Username  string `json:"username"`
+	Name      string `json:"name"`
+	State     string `json:"state"`
+	Locked    bool   `json:"locked"`
+	AvatarURL string `json:"avatar_url"`
+	WebURL    string `json:"web_url"`
+}
+
+type TimeStats struct {
+	TimeEstimate        *int    `json:"time_estimate"`
+	TotalTimeSpent      *int    `json:"total_time_spent"`
+	HumanTimeEstimate   *string `json:"human_time_estimate"`
+	HumanTotalTimeSpent *string `json:"human_total_time_spent"`
+}
+
+type TaskCompletionStatus struct {
+	Count          int `json:"count"`
+	CompletedCount int `json:"completed_count"`
+}
+
+type Links struct {
+	Self                string  `json:"self"`
+	Notes               string  `json:"notes"`
+	AwardEmoji          string  `json:"award_emoji"`
+	Project             string  `json:"project"`
+	ClosedAsDuplicateOf *string `json:"closed_as_duplicate_of"`
+}
+
+type References struct {
+	Short    string `json:"short"`
+	Relative string `json:"relative"`
+	Full     string `json:"full"`
+}
+
+func GetOAuthConfig() (*oauth2.Config, []oauth2.AuthCodeOption, error) {
+	var (
+		ok                 bool
+		gitlabClientID     string
+		gitlabClientSecret string
+		frontendUri        string
+	)
+	if gitlabClientID, ok = os.LookupEnv("GITLAB_CLIENT_ID"); !ok || gitlabClientID == "" {
+		return nil, nil, errors.New("GITLAB_CLIENT_ID not set")
+	}
+	if gitlabClientSecret, ok = os.LookupEnv("GITLAB_CLIENT_SECRET"); !ok || gitlabClientSecret == "" {
+		return nil, nil, errors.New("GITLAB_CLIENT_SECRET not set")
+	}
+	if frontendUri, ok = os.LookupEnv("REACT_APP_FRONTEND_URI"); !ok || frontendUri == "" {
+		return nil, nil, errors.New("REACT_APP_FRONTEND_URI not set")
+	}
+
+	options := []oauth2.AuthCodeOption{
+		oauth2.AccessTypeOffline,
+	}
+
+	return &oauth2.Config{
+		ClientID:     gitlabClientID,
+		ClientSecret: gitlabClientSecret,
+		Endpoint:     gitlabEndpoint,
+		RedirectURL:  fmt.Sprintf("%s/callback/gitlab", frontendUri),
+	}, options, nil
+}
+
+func GetAccessToken(ctx context.Context, code string) (*oauth2.Token, error) {
+	conf, opts, err := GetOAuthConfig()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return conf.Exchange(ctx, code, opts...)
+}
+
+func doGitlabPostRequest[TOut any, TIn any](accessToken string, url string, input TIn) (TOut, error) {
+	var zero TOut
+	b, err := json.Marshal(input)
+	if err != nil {
+		return zero, err
+	}
+
+	return doGitlabRequest[TOut]("POST", accessToken, url, string(b))
+}
+
+func doGitlabGetRequest[T any](accessToken string, url string) (T, error) {
+	return doGitlabRequest[T]("GET", accessToken, url, "")
```

Comment:
Looks good - not sure the rate limit here, but just want to call out that a future improvement can be to cache these results

Reply:
Sure.